### ISSUE TITLE
tap or junit output

### ISSUE THREAD

**Issue by linuxmaniac** [mpeterv/luacheck#19]
It would be nice to be able to get the results in TAP or JUnit format in order to integrate luacheck with jenkins jobs.

For a good reference for TAP format, check: http://testanything.org/tap-specification.html
For more junit format information, check: https://svn.jenkins-ci.org/trunk/hudson/dtkit/dtkit-format/dtkit-junit-model/src/main/resources/com/thalesgroup/dtkit/junit/model/xsd/junit-4.xsd

**Comment by mpeterv** [comment]
I agree, I'll look to implement TAP and JUnit formatters as well as an option to use some other custom formatter. I'm not sure what should be the relationship between testsuites and tests and checked files and warnings. E.g. in TAP I could make each warning a separate test:

```
1..3
ok 1 src/foo.lua
not ok 2 src/bar.lua:3:5: accessing undefined variable 'baz'
not ok 3 src/bar.lua:5:8: unused variable 'var'
```

.. or I could make each file correspond to a test and output warnings as diagnostics:

```
1..2
ok 1 src/foo.lua
not ok 2 src/bar.lua
# src/bar.lua:3:5: accessing undefined variable 'baz'
# src/bar.lua:5:8: unused variable 'var'
```

.. which has the advantage of number of tests being constant (same as number of checked files), but relies on test harness to interpret warning messages properly.

Similarly, for JUnit I could make each file correspond to a testsuite and each warning to a failed test case, or I could produce one testsuite that contains a testcase per file, with warning info contained in `system-out` or `system-err` of testcases. What do you think?

**Comment by linuxmaniac** [comment]
I would go for "make each warning a separate test" for TAP and "make each file correspond to a testsuite and each warning to a failed test case" for JUnit

**Comment by mpeterv** [comment]
I have implemented TAP formatter (plus a plain formatter for piping `luacheck` output into `grep`, etc; there is also possibility to use custom formatters) on master branch. Formatter can be selected using `--formatter` option, e.g. `--formatter=TAP`. You can install from master branch using luarocks by running `sudo luarocks install  --server=http://rocks.moonscript.org/manifests/mpeterv luacheck scm`.

**Comment by linuxmaniac** [comment]
Thanks! I'll check it ASAP
BTW, JFYI https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=776473

**Comment by mpeterv** [comment]
I've implemented JUnit formatter, it can be selected using `--formatter=JUnit`. I went for testcase-per-file approach, I figured using `<failure>` nodes to represent warnings would make sense. Example output:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="Luacheck report" tests="3">
    <testcase name="spec/samples/good_code.lua" classname="spec/samples/good_code.lua"/>
    <testcase name="spec/samples/bad_code.lua" classname="spec/samples/bad_code.lua">
        <failure type="W211" message="spec/samples/bad_code.lua:3:16: unused function 'helper'"/>
        <failure type="W212" message="spec/samples/bad_code.lua:3:23: unused variable length argument"/>
        <failure type="W111" message="spec/samples/bad_code.lua:7:10: setting non-standard global variable 'embrace'"/>
        <failure type="W412" message="spec/samples/bad_code.lua:8:10: variable 'opt' was previously defined as an argument on line 7"/>
        <failure type="W113" message="spec/samples/bad_code.lua:9:11: accessing undefined variable 'hepler'"/>
    </testcase>
    <testcase name="spec/samples/python_code.lua" classname="spec/samples/python_code.lua">
        <error type="Syntax error"/>
    </testcase>
</testsuite>
```

Thanks for the info, is there any cooperation needed w.r.t the ITP?

**Comment by linuxmaniac** [comment]
> Thanks for the info, is there any cooperation needed w.r.t the ITP?
I will ask you directly by email about ITP But I would like to package a release with this included.

is 0.9.0 feasible?

**Comment by mpeterv** [comment]
> is 0.9.0 feasible?

It is. I planned to release it rather soon, somewhere around this Friday.

**Comment by mpeterv** [comment]
0.9.0 released.
